### DISCUSSION TITLE
Help menu should explain the meaning of flags

### DISCUSSION THREAD

**Discussion by whitney-cumber** [neomutt/neomutt#4525] [Development]
## What would you like to be able to do?

Quickly look up the meaning of character flags in the help menu.

NeoMutt uses a lot a single character indicators (and even allows to customise some).  Keeping track of all the meanings is a burden, especially for menus you seldom use.  Some examples for such "character flags" are:

* `%t` validity in `$gpg_entry_format`
* `%f` flags in `$alias_format`
* `%c` charset conversion in `$attach_format`
* `%t` in tagged in `$folder_format`
* `%S` status in `$index_format`
* `%T` to chars in `$index_format` (the values can be configured in `$to_chars`)
* ...

(There might even be some character flags which are not produced by expandos, though I haven't looked for them explicitly)

It would be really nice if NeoMutt would print an explanation for these flags in the help menu (usually accessed via `?`).  That way a user can quickly lookup what a flag means.  This is particularly useful for flags which you do not encounter often.

This would also lower the bar of entrance for using NeoMutt in the first place.

**Comment by flatcap** [comment]
> Quickly look up the meaning of character flags in the help menu.
> This would also lower the bar of entrance for using NeoMutt in the first place.

That all sounds good, and there's a precedent for this sort of thing.
If you `<limit>` in the index, then type <kbd>~</kbd><kbd>\<Tab\></kbd>, you'll get a help page of patterns.

What I can't quite figure out, is where you'd like the help to be "anchored".
What would the user being doing?
What would they hit to activate the help?

**Comment by whitney-cumber** [comment]
> If you <limit> in the index, then type ~<Tab>, you'll get a help page of patterns.

Didn't know that thanks!  (Previously I always had to open `man neomuttrc` to look up patters I use only from time to time.)

>What I can't quite figure out, is where you'd like the help to be "anchored".
>What would the user being doing?
>What would they hit to activate the help?

Simple: press `?`  (or trigger the `<help>` function in any other way)

Currently, the help menu/screen only shows only key bindings (so is more of a keybinding menu).
My thought was to add another section to the end, listing the flags applicable to the current menu (e.g. index, or attachment), like this:

```
N – new message
O – old message
* – tagged message
...
```

---

Another idea which jumped into my mind when looking at `htop`'s help screen is this:
Render one (multiple?) sample entries with explanations, e.g.
```
 123 rP  Dec 13 Polar Fox       ( 23K) Happy Winter
   ^ ^^^ ^    ^ ^                   ^  ^
   | ||| |    | |                   |  |
   a bcd e    f g                   h  i

a – message number
b – status flags: N = new mail, n = new mail in this thread, O = old mail, o = old mail in this thread, r = replied to mail
c – other flags: s = signed but not verified, S = verified, P = encrypted, D = marked for deletion, ...
d – `*` if message is tagged, blank otherwise
e – month of date sent
f – day of date sent
g – sender
h – messages size
i – subject
```

This is quite advanced and potentially difficult because of the flexibility of the format.  Figuring out the position seems easy unless you account that we have conditional expandos `%n?%d&%a?`  (yes that's a stupid thing but possible).

But maybe if we ignore such conditionals in the explanations, that's at least a start of explaining most cases (and probably those which occur often)

**Comment by flatcap** [comment]
Sounds good.

If we added a new `<help-context>` function to the `generic` and `editor` menus, then it'd be available everywhere.
Being a _function_, would mean it would get handled by the focussed dialog -- each providing context-specific help.

A good start would be:
- "Official" name of that dialog, e.g. "Index"
- A list of config items that apply to _that_ dialog
- A glossary of terms relating to _that_ dialog 
- Links to the web docs

**Comment by g0mb4** [comment]
I think #3925 partly solves this problem.  An explanation for `$index_format` could be useful for the help menu of index.
